File: src/schemas/image.ts
import { z } from 'zod';

export const ImageTypeSchema = z.enum(['File', 'Slice']);
export type ImageType = z.infer<typeof ImageTypeSchema>;

export const ImageSchema = z.object({
  hash: z.string(),
  image_name: z.string().nullable(),
  width: z.number(),
  height: z.number(),
  image_type: ImageTypeSchema,
  volume_id: z.string().nullable(),
  slice_index: z.number().nullable(),
  mask_hash: z.string().nullable(),
});
export type Image = z.infer<typeof ImageSchema>;

export const ImageUpsertInputSchema = z.object({
  hash: z.string(),
  image_name: z.string().nullable().optional(),
  width: z.number(),
  height: z.number(),
  image_type: ImageTypeSchema,
  volume_id: z.string().nullable().optional(),
  slice_index: z.number().nullable().optional(),
  mask_hash: z.string().nullable().optional(),
});
export type ImageUpsertInput = z.infer<typeof ImageUpsertInputSchema>;

File: crates/db/src/models/image.rs
use crate::models::annotation::Annotation;

#[derive(Debug, PartialEq, toasty::Embed, serde::Serialize, serde::Deserialize)]
pub enum ImageType {
    File,
    Slice,
}

#[derive(Debug, toasty::Model, serde::Serialize)]
pub struct Image {
    #[key]
    pub hash: String,

    pub image_name: Option<String>,

    pub width: i32,

    pub height: i32,

    pub image_type: ImageType,

    #[index]
    pub volume_id: Option<String>,

    pub slice_index: Option<i32>,

    pub mask_hash: Option<String>,

    #[serde(skip)]
    #[has_many]
    pub annotations: Vec<Annotation>,
}
